src/check/context/clss/concrete/variant.rs

```diff
@@ -0,0 +1,167 @@
+            }
+            other => {
+                let msg = format!("{} cannot have parent", other);
+                Err(vec![TypeErr::new(pos, &msg)])
+            }
+        }
+    }
+}
+
+impl HasParent<&TrueName> for ClassVariant {
+    fn has_parent(&self, name: &TrueName, ctx: &Context, pos: Position) -> TypeResult<bool> {
+        match &self {
+            ClassVariant::Direct(union) => {
+                Ok(union.iter().all(|class| class.has_parent(name, ctx, pos).is_ok()))
+            }
+            other => {
+                let msg = format!("{} cannot have parent", other);
+                Err(vec![TypeErr::new(pos, &msg)])
+            }
+        }
+    }
+}
+
+impl Hash for ClassVariant {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        match &self {
+            ClassVariant::Direct(class_set) => {
+                class_set.iter().sorted_by_key(|c| &c.name).for_each(|c| c.hash(state))
+            }
+            ClassVariant::Tuple(classes) => classes.hash(state),
+            ClassVariant::Fun(args, ret) => {
+                args.hash(state);
+                ret.hash(state)
+            }
+        };
+    }
+}
+
+impl PartialEq for ClassVariant {
+    fn eq(&self, other: &Self) -> bool {
+        match (&self, other) {
+            (ClassVariant::Direct(l_set), ClassVariant::Direct(r_set)) => {
+                let l_set: Vec<&Class> = l_set.iter().sorted_by_key(|c| &c.name).collect();
+                let r_set: Vec<&Class> = r_set.iter().sorted_by_key(|c| &c.name).collect();
+                l_set == r_set
+            }
+            (ClassVariant::Tuple(l_c), ClassVariant::Tuple(r_c)) => l_c == r_c,
+            (ClassVariant::Fun(l_a, l_r), ClassVariant::Fun(r_a, r_r)) => l_a == r_a && l_r == r_r,
+            _ => false,
+        }
+    }
+}
+
+impl GetField<FieldUnion> for ClassVariant {
+    fn field(&self, name: &str, ctx: &Context, pos: Position) -> TypeResult<FieldUnion> {
+        match &self {
+            ClassVariant::Direct(class_set) => {
+                let fields: HashSet<Field> = class_set
+                    .iter()
+                    .map(|class| class.field(name, ctx, pos))
+                    .collect::<Result<_, _>>()?;
+                Ok(FieldUnion::from(&fields))
+            }
+            other => {
+                let msg = format!("'{}' cannot define a field", other);
+                Err(vec![TypeErr::new(pos, &msg)])
+            }
+        }
+    }
+}
+
+impl GetFun<FunUnion> for ClassVariant {
+    fn fun(&self, name: &StringName, ctx: &Context, pos: Position) -> TypeResult<FunUnion> {
+        match &self {
+            ClassVariant::Direct(class_set) => {
+                let funs: HashSet<Function> = class_set
+                    .iter()
+                    .map(|class| class.fun(name, ctx, pos))
+                    .collect::<Result<_, _>>()?;
+                Ok(FunUnion::from(&funs))
+            }
+            other => {
```

Comment:
We can do this more elegantly by treating `ClassVariant::Tuple` as a `tuple` class, for which Python does have some methods:
`index`, `count`, `__str__`, etc.

So we just call that class and treat this as though one called `ClassVariant::Direct(class_set = {Tuple})`

Reply:
Same for `ClassVariant::Fun`, python already defines some methods for this class, such as `__str__`